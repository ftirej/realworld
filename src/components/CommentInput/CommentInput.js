import React, { useState } from "react";
import { makeStyles } from "@material-ui/core/styles";
import Card from "@material-ui/core/Card";
import CardActions from "@material-ui/core/CardActions";
import CardContent from "@material-ui/core/CardContent";
import Avatar from "@material-ui/core/Avatar";
import TextareaAutosize from "@material-ui/core/TextareaAutosize";
import Button from "@material-ui/core/Button";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import * as feed from "../../store/feed/actions";

const useStyles = makeStyles({
  card: {
    minWidth: 275,
    marginTop: "10px"
  },
  title: {
    fontSize: 14
  },
  pos: {
    marginBottom: 12
  }
});

const initialState = {
  comment: { body: "" }
};

const CommentInput = props => {
  const classes = useStyles();
  const [formData, setFormData] = useState(initialState);

  const handleSubmitPost = event => {
    event.preventDefault();
    props.feed.postCommentList(props.slugId, formData);
  };

  const handleBodyChange = event => {
    setFormData({ comment: { body: event.target.value } });
  };

  return (
    <form onSubmit={handleSubmitPost}>
      <Card className={classes.card} variant="outlined">
        <CardContent style={{ padding: "0" }}>
          <TextareaAutosize
            style={{
              width: "100%",
              borderStyle: "hidden",
              padding: "1.25rem",
              outline: "none"
            }}
            className={classes.MuiTypography}
            aria-label="minimum height"
            rowsMin={4}
            placeholder="Write a comment..."
            value={formData.body}
            onChange={handleBodyChange}
          />
        </CardContent>
        <CardActions
          style={{
            backgroundColor: "#f5f5f5",
            display: "flex",
            justifyContent: "space-between"
          }}
        >
          <div>
            <Avatar
              aria-label="recipe"
              className={classes.avatar}
              src={props.article.author.image}
            ></Avatar>
          </div>
          <Button
            size="small"
            type="submit"
            color="primary"
            style={{
              color: "#fff",
              backgroundColor: "#5cb85c",
              borderColor: "#5cb85c"
            }}
          >
            Post Comment
          </Button>
        </CardActions>
      </Card>
    </form>
  );
};

const mapDispatchToProps = dispatch => {
  return {
    feed: bindActionCreators({ ...feed }, dispatch)
  };
};

export default connect(null, mapDispatchToProps)(CommentInput);
