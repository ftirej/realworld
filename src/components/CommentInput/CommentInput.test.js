import React from "react";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { shallow } from "enzyme";
import { CommentInput } from "./CommentInput";
import Card from "@material-ui/core/Card";
import CardActions from "@material-ui/core/CardActions";
import CardContent from "@material-ui/core/CardContent";
import Avatar from "@material-ui/core/Avatar";
import TextareaAutosize from "@material-ui/core/TextareaAutosize";
import Button from "@material-ui/core/Button";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const setup = (initalState = {}, props = {}) => {
  const store = mockStore(initalState);
  const wrapper = shallow(<CommentInput store={store} {...props} />);
  // console.log(wrapper.debug());
  return wrapper;
};

describe("CommentInput", () => {
  describe("WHEN rendering", () => {
    let wrapper;

    const baseProps = {
      article: {
        author: {
          image: "https://i.stack.imgur.com/xHWG8.jpg"
        }
      }
    };

    beforeEach(() => {
      wrapper = setup({}, baseProps);
    });

    it("should render an Input Card", () => {
      expect(wrapper.find(Card)).toHaveLength(1);
      expect(wrapper.find(CardActions)).toHaveLength(1);
      expect(wrapper.find(CardContent)).toHaveLength(1);
      expect(wrapper.find(CardContent)).toHaveLength(1);
      expect(wrapper.find(Avatar)).toHaveLength(1);
      expect(wrapper.find(TextareaAutosize)).toHaveLength(1);
      expect(wrapper.find(Button)).toHaveLength(1);
    });

    it("should render a button as a submit type", () => {
      expect(wrapper.find(Button).props().type).toBe("submit");
    });
  });

  describe("when submit button is clicked", () => {
    let wrapper;
    const mockFunc = jest.fn();

    const baseProps = {
      article: {
        author: {
          image: "https://i.stack.imgur.com/xHWG8.jpg"
        }
      },
      feed: {
        postCommentList: mockFunc
      }
    };

    beforeEach(() => {
      wrapper = setup({}, baseProps);
    });

    it("should render with given state from Redux store", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should dispatch an action on button click", () => {
      wrapper.find("form").simulate("submit", { preventDefault: jest.fn() });
      const callback = mockFunc.mock.calls.length;
      expect(callback).toBe(1);
    });
  });
});
