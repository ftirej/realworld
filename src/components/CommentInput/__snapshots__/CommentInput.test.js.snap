// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentInput when submit button is clicked should render with given state from Redux store 1`] = `ShallowWrapper {}`;
